components/booking/TimeOptions.tsx

```diff
@@ -1,48 +1,56 @@
 // TODO: replace headlessui with radix-ui
 import { Switch } from "@headlessui/react";
+import { useRouter } from "next/router";
 import { FC, useEffect, useState } from "react";
 import TimezoneSelect, { ITimezoneOption } from "react-timezone-select";
 
 import classNames from "@lib/classNames";
+import { timeZone } from "@lib/clock";
 import { useLocale } from "@lib/hooks/useLocale";
-
-import { is24h, timeZone } from "../../lib/clock";
+import { useToggleQuery } from "@lib/hooks/useToggleQuery";
 
 type Props = {
   localeProp: string;
   onSelectTimeZone: (selectedTimeZone: string) => void;
-  onToggle24hClock: (is24hClock: boolean) => void;
 };
 
 const TimeOptions: FC<Props> = (props) => {
   const [selectedTimeZone, setSelectedTimeZone] = useState("");
   const [is24hClock, setIs24hClock] = useState(false);
   const { t } = useLocale({ localeProp: props.localeProp });
+  const hour12 = useToggleQuery("hour12", { zero: true });
+  const router = useRouter();
 
   useEffect(() => {
-    setIs24hClock(is24h());
+    setIs24hClock(
+      (router.query.hour12 !== undefined && !hour12.isOn) ||
+        !new Intl.DateTimeFormat(props.localeProp, { timeStyle: "short" })
+          .format(new Date())
+          .match(/\d+:\d+ (AM|PM)/)
+    );
     setSelectedTimeZone(timeZone());
-  }, []);
-
-  useEffect(() => {
-    if (selectedTimeZone && timeZone() && selectedTimeZone !== timeZone()) {
-      props.onSelectTimeZone(timeZone(selectedTimeZone));
-    }
-  }, [selectedTimeZone]);
+  }, [props.localeProp, hour12.isOn, router.query.hour12]);
 
   const handle24hClockToggle = (is24hClock: boolean) => {
     setIs24hClock(is24hClock);
-    props.onToggle24hClock(is24h(is24hClock));
+    router.push(is24hClock ? hour12.hrefOff : hour12.hrefOn);
+  };
```

Comment:
Also @emrysal I couldn't toggle between 12/24 hours times in here IDK why.

Reply:
It's stays as 24 hours always.

Reply:
This is really odd, have a look whether this happens for you @ https://calendso-git-feature-i18n-datetimeformat-builtin-cal.vercel.app/bailey/demo

Reply:
https://user-images.githubusercontent.com/3504472/136708527-d801dc70-75cd-4515-8e17-378c222be2e7.MOV

Yup